File: backend/app/connectors/onedrive.py
"""Conector de OneDrive / Microsoft Graph (importación puntual).

El frontend autentica con MSAL (popup, scope Files.Read), el usuario elige
archivos y acá los descargamos vía Microsoft Graph. Los archivos de OneDrive son
binarios reales (pdf/docx/...) → descarga directa, sin export como en Google.
"""
from __future__ import annotations

from pathlib import Path

import httpx

from app.ingestion.parsers import SUPPORTED_EXTENSIONS

SOURCE_ONEDRIVE = "onedrive"

_GRAPH = "https://graph.microsoft.com/v1.0"


async def download_files(
    access_token: str, files: list[dict]
) -> tuple[list[tuple[str, bytes, str | None]], list[str], list[str]]:
    """Descarga los archivos elegidos por id de Graph.

    Devuelve (items, rechazados, fallidos)."""
    items: list[tuple[str, bytes, str | None]] = []
    rejected: list[str] = []
    failed: list[str] = []
    headers = {"Authorization": f"Bearer {access_token}"}

    # follow_redirects: el endpoint /content responde 302 a una URL prefirmada.
    async with httpx.AsyncClient(timeout=60, follow_redirects=True) as client:
        for f in files:
            fid = f.get("id")
            name = f.get("name") or (fid or "documento")
            if not fid:
                failed.append(name)
                continue
            if Path(name).suffix.lower() not in SUPPORTED_EXTENSIONS:
                rejected.append(name)
                continue
            try:
                resp = await client.get(
                    f"{_GRAPH}/me/drive/items/{fid}/content", headers=headers
                )
                resp.raise_for_status()
                ctype = resp.headers.get("content-type")
                items.append((name, resp.content, ctype))
            except Exception:  # noqa: BLE001
                failed.append(name)

    return items, rejected, failed

File: frontend/src/lib/oneDrive.ts
/**
 * Integración con OneDrive / Microsoft Graph (importación puntual).
 * MSAL (popup, scope Files.Read) → access_token → exploramos el drive con Graph
 * y el backend descarga los archivos elegidos.
 *
 * Config: Client ID de una App registration de Azure (SPA, redirect =
 * window.location.origin). Se setea desde la UI (localStorage) o por env
 * (VITE_MS_CLIENT_ID). No necesita "API key" (Graph usa solo el token).
 */
import { PublicClientApplication } from "@azure/msal-browser";

const env = (import.meta as unknown as { env: Record<string, string | undefined> })
  .env;
const LS_CLIENT = "aidoc-ms-client-id";
const GRAPH = "https://graph.microsoft.com/v1.0";
const SCOPES = ["Files.Read"];

export interface DriveItem {
  id: string;
  name: string;
  isFolder: boolean;
  size?: number;
}

function ls(key: string): string | undefined {
  try {
    return localStorage.getItem(key) || undefined;
  } catch {
    return undefined;
  }
}

export function getOneDriveClientId(): string | undefined {
  return ls(LS_CLIENT) || env.VITE_MS_CLIENT_ID || undefined;
}

export function setOneDriveClientId(clientId: string): void {
  try {
    localStorage.setItem(LS_CLIENT, clientId.trim());
  } catch {
    /* ignore */
  }
}

export function isOneDriveConfigured(): boolean {
  return Boolean(getOneDriveClientId());
}

let _pca: PublicClientApplication | null = null;
let _initClientId: string | null = null;

async function getPca(): Promise<PublicClientApplication> {
  const clientId = getOneDriveClientId();
  if (!clientId) throw new Error("OneDrive no está configurado.");
  if (!_pca || _initClientId !== clientId) {
    _pca = new PublicClientApplication({
      auth: {
        clientId,
        authority: "https://login.microsoftonline.com/common",
        redirectUri: window.location.origin,
      },
      cache: { cacheLocation: "localStorage" },
    });
    await _pca.initialize();
    _initClientId = clientId;
  }
  return _pca;
}

/** Intenta un access_token SIN popup (devuelve null si requiere interacción). */
export async function getOneDriveTokenSilent(): Promise<string | null> {
  const pca = await getPca();
  const accounts = pca.getAllAccounts();
  if (accounts.length === 0) return null;
  try {
    const res = await pca.acquireTokenSilent({
      scopes: SCOPES,
      account: accounts[0],
    });
    return res.accessToken;
  } catch {
    return null;
  }
}

/** Obtiene un access_token con popup. DEBE llamarse desde un gesto del usuario
 * (clic): los navegadores bloquean popups que no nacen de una interacción. */
export async function getOneDriveToken(): Promise<string> {
  const silent = await getOneDriveTokenSilent();
  if (silent) return silent;
  const pca = await getPca();
  const res = await pca.acquireTokenPopup({ scopes: SCOPES });
  return res.accessToken;
}

/** Lista el contenido de una carpeta (root si no se pasa id). */
export async function listChildren(
  token: string,
  folderId?: string
): Promise<DriveItem[]> {
  const path = folderId
    ? `${GRAPH}/me/drive/items/${folderId}/children`
    : `${GRAPH}/me/drive/root/children`;
  const url = `${path}?$select=id,name,folder,file,size&$top=200`;
  const resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!resp.ok) throw new Error(`Graph ${resp.status}`);
  const data = await resp.json();
  return (data.value || []).map(
    (it: { id: string; name: string; folder?: unknown; size?: number }) => ({
      id: it.id,
      name: it.name,
      isFolder: Boolean(it.folder),
      size: it.size,
    })
  );
}
